### ISSUE TITLE
Zinc doesn't like type aliases + type projections

### ISSUE THREAD

**Issue by jvican** [sbt/zinc#535]
@romanowski submitted a PR with an example that fails to compile in Zinc in https://github.com/sbt/zinc/pull/475. At first, I thought that the error came from the fact that we were not detecting an API change in the type alias. This was a result of misreading the debug logs.
  
The actual error happening here has more to do with scalac, and the breaking of one invariant of Zinc: given a change, Zinc can always compile the invalidated changes and then detect further changes based on the incremental compiler analysis. This doesn't hold for some the corner case I explain next.
  
The issue at hand is that `nested-type-params` fails to compile the invalidated changes because Scalac's aggressive type dealiasing. The error only happens when type aliases are used in type projections. The initial example is:

```scala
// A.scala
trait A { def a = 1 }
// B.scala
trait B { def b = 1.0 }
// Provider.scala
trait Provider { type Operations = A }
// Foo.scala
object Foo { def provide: Provider#Operations }
// Bar.scala
object Bar { Foo.provide.a }
```

When we change, `Provider` to `trait Provider { type Operations = B }` and `Bar` to `object Bar { Foo.provide.b }`, scalac tries to compile and fail. The failure comes from the fact that `Foo.provide` has type `A`. The type `A` comes from the pickle in the class file and it's generated by the first full compilation.
  
Ideally, Zinc should see that `Foo.provide` has type `Provider#Operations` instead of `A`. It only sees `A` because the compiler aggressively dealiases the type projection to `A`. However, this only seems to happen for the case where `Provider` is a type.
  
This hypothesis is confirmed by the fact that this bug cannot be reproduced if `Operations` is instead a class extending `A` and afterwards extending `B`, or if `Provider` is an object (term). Also, making `Provider` a class reproduces the issue.
  
###  Solution

The obvious solution is to fix this in the compiler, so that the pickles contain all the information. However, that isn't doable, since this is more or less a fundamental limitation.

Therefore, we're left with the need for a fix in Zinc. The fix I propose is that we introduce a new kind of dependency (or reuse an existing one, it's not yet clear) so that when we traverse `Foo.provide.a` we're able to register the `Foo` dependency on Bar, and be able to add that dependency to the initial invalidated sources in `IncrementalCommon`. This kind of dependency is new in that it's the first kind of dependency that invalidates a source file (of the same project) in the first iteration of the incremental compile.

This solution will need to re-use the infrastructure introduced in https://github.com/sbt/zinc/pull/497, and it will only be present in 2.12.5+. 
  
I still need to double-check that the implementation is doable, but it seems we're on the right track.

**Comment by smarter** [comment]
>  It only sees A because the compiler aggressively dealiases the type projection to A

That may be the case, but have you double-checked that Zinc is not the one doing the dealiasing here? I'm thinking of this: https://github.com/sbt/zinc/blob/e245d953127627ee5127c846c069c33f7d9c3c2a/internal/compiler-bridge/src/main/scala/xsbt/ExtractAPI.scala#L469-L472

**Comment by jvican** [comment]
@smarter Yeah, pretty sure that's not the source of our error since even if `ExtractAPI` was not dealiasing `Provider#Operations` Zinc will fail because internal changes via `ExtractAPI` are not considered for the first iteration of the incremental compiler (when all the modified sources are compiled).

**Comment by smarter** [comment]
> Zinc will fail because internal changes via ExtractAPI are not considered for the first iteration of the incremental compiler (when all the modified sources are compiled).

I'm not exactly sure I understand the issue here, but note that the corresponding test case already works fine in dotty: https://github.com/lampepfl/dotty/blob/master/sbt-dotty/sbt-test/source-dependencies/nested-type-params/test

**Comment by jvican** [comment]
That's because you don't constant fold in typer, but in post typer, right? The issue here is that the aggressive inlining of constants prevents the first compile iteration from succeeding. If you have a look at the original code it will be clearer.

**Comment by smarter** [comment]
We constant fold in FirstTransform after pickler, but I don't see how constant folding is involved in the example you give, do you have a link to the "original code" you're refering to?

**Comment by jvican** [comment]
> We constant fold in FirstTransform after pickler,

This is the reason why it doesn't happen then. The aggressive constant folding in typer makes pickler in scalac not preserve the information required for the following paragraph to behave correctly:

> Ideally, Zinc should see that Foo.provide has type Provider#Operations instead of A. It only sees A because the compiler aggressively dealiases the type projection to A. However, this only seems to happen for the case where Provider is a type.
  
As Dotty doesn't constant fold that sooner, `Foo.provide` points to `Provider#Operations` instead of `A`.

**Comment by smarter** [comment]
OK, I understand what you mean now, it's just that we're using different vocabulary here. In the compiler, we *dealias* types but we *constant fold* terms and those are completely unrelated operations (but Dotty tries to not do either of these things before pickling).

**Comment by jvican** [comment]
You're right, I meant dealiasing and not constant folding in my last comments. I just had the latter word fresh in my mind. But I do mention dealiasing in the description of the ticket in "because the compiler aggressively dealiases the type projection to A". Great to see we're on the same page now.

**Comment by jvican** [comment]
Blocked until 2.13, when we'll merge https://github.com/sbt/zinc/pull/497 and I'll submit a solution for this based on that PR.

**Comment by Friendseeker** [comment]
#475 is passing after #1284

Since the cause of the issue, as mentioned, is:

> The actual error happening here has more to do with scalac, and the breaking of one invariant of Zinc: given a change, Zinc can always compile the invalidated changes and then detect further changes based on the incremental compiler analysis. This doesn't hold for some the corner case I explain next.

I don't have a full grasp of the issue, but I guess #1284 probably fixed it.

**Comment by jozanek** [comment]
Same reproduction as #476 (`nested-type-params` from #475), fixed by #1780. #1284 also fixed it but was reverted by #1462, so the comment above is outdated.

The dealiasing is scalac's, not `ExtractAPI`'s: an alias selected through an unstable prefix goes through `asSeenFrom`, which captures `this` as an existential and runs `normalizeAliases` in `existentialAbstraction`. So `Provider#Operations` is pickled as `A` (checked with `-Xprint:typer` on 2.13.18), while `Provider.Operations` on an object stays intact. Scala 3 keeps the projection.

On its own that is harmless: changing only `Providers.scala` recompiles Provider, Foo, Bar in three rounds. It breaks when `Bar.scala` changes in the same round and compiles against Foo's stale pickle. That is what #1780 retries, so this can close with it.
